Use app namespace for framework tasks

ryohashimoto opened the pull request:
By introducing Rake proxy, `rake rails` commands become `rails rails` commands (`rails rails:template`, `rails rails:update` etc.) .

For simplicity, I removed the `rails` namespace from the framework codes and docs. (and now we can use `rails template`, `rails update` etc.)

---

updated:

Not removing the rails namespace, use app namespace instead.
e.g. `rails rails:template` and `rails rails:update` tasks should be `rails app:template`, `rails app:update`

ryohashimoto commented on railties/lib/rails/tasks/framework.rake:
```diff
@@ -1,68 +1,66 @@
-namespace :rails do
```
Indent only changes on this file.

kaspth commented:
I don't think we should change this right now. I think more of `bin/rails` as the runner, and that the commands it runs pertains to the app itself... mostly. Because some commands fiddle more with the framework, i.e. those in the `rails` namespace. I say we keep it like this for 5.0 and spare the deprecation cycle. Thanks!

matthewd commented:
@kaspth should we change the namespace to `app:` or something? `rails rails:foo` does look pretty terrible.

ryohashimoto commented:
@matthewd 
I think so too.
The `rails` namespace is used for multiple purposes (`upgrade` and `template`). So we could remove the namespace.

kaspth commented:
Yeah, app sounds good.

We should still define the rails versions but enrich them with a deprecation prerequisite task.

ryohashimoto commented:
@kaspth 
Thanks for reopening this PR.
I updated the PR to use `app` namespace and add deprecation warnings for existing tasks under `rails` namespace.

kaspth commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+* Use app namespace instead of rails namespace from the framework rake tasks.
```
Needs more info. Mention some of the tasks and that the `rails:` namespace has been deprecated.

kaspth commented on railties/lib/rails/tasks/framework.rake:
```diff
@@ -66,3 +66,23 @@ namespace :rails do
+  task :deprecated do
+    Rake.application.top_level_tasks.grep(/^rails:/).each do |task|
+      $stderr.puts "WARNING: #{task} is deprecated. Please use #{task.sub(/^rails:/, 'app:')} instead."
+    end
```
Don't need to print deprecation warnings for every task in the namespace. Likely more confusing than just printing the task they ran, if we can.

The message should mention that it's the namespace that is deprecated. Going with something like:

```Running this task with the rails task namespace is deprecated in favor of app. Please run e.g. bin/rails app:update instead.
```

Could even remove the need to print the task altogether.

We should also require `active_support` and use `ActiveSupport::Deprecation.warn`, so people get their configured deprecation behavior to fire.

kaspth commented:
Also needs a rebase (and remember to keep your commits squashed). Thanks :heart:

ryohashimoto committed:
The tasks in the rails task namespace is deprecated in favor of app namespace.
(e.g. `rails:update` and `rails:template` tasks is renamed to `app:update` and `app:template`.)

ryohashimoto commented:
@kaspth
Thanks for your advice.
Modified to use `ActiveSupport::Deprecation.warn` and removed `deprecated` task in the rails namespace.
So, when the task in the rails namespace is executed, only a deprecation message is printed out.
Is it correct?

kaspth commented on railties/lib/rails/tasks/framework.rake:
```diff
@@ -66,3 +68,24 @@ namespace :rails do
+      Running this task with the rails task namespace is deprecated in favor of app.
+      Please run e.g. bin/rails #{task.sub(/^rails:/, 'app:')} instead."
+    MSG
+  end
```
Sorry wasn't clear enough.

> Don't need to print deprecation warnings for every task in the namespace.

I meant we shouldn't use `Rake.application.top_level_tasks.grep(/^rails:/)`. I didn't mean to remove the deprecation enrichment of the `rails:` namespaced tasks, that part was correct before.

kaspth replied:
We should replace this with:

```ruby
task :deprecated do |task|
  ActiveSupport::Deprecation.warn(<<-MSG.squish)
   Running #{task.name.sub('deprecated:', '')} with the rails: namespace is deprecated in favor of app.
   Run e.g. bin/rails app:update instead."
 MSG
end
```

ryohashimoto replied:
I modified the codes for the :rails namespace in my local environment.

```ruby
namespace :rails do
  task :deprecated do |task|
    ActiveSupport::Deprecation.warn(<<-MSG.squish)
      Running #{task.name.sub('deprecated:', '')} with the rails: namespace is deprecated in favor of app.
      Run e.g. bin/rails app:update instead."
    MSG
  end

  task :update => ['deprecated', 'app:update']
  task :template => ['deprecated',  'app:template']

  namespace :templates do
    task :copy => ['deprecated', 'app:templates:copy']
  end

  namespace :update do
    task :configs => ['deprecated', 'app:update:configs']
    task :bin => ['deprecated', 'app:update:bin']
  end
end
```

and ran the command.

```
$ rails rails:update                                                                  1 ↵  
DEPRECATION WARNING: Running rails:deprecated with the rails: namespace is deprecated in favor of app. Run e.g. bin/rails app:update instead.". (called from block in execute at /Users/ryo/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:238)
```

The message is incorrect because `rails:deprecated` is the original task used by the actual command.

ryohashimoto replied:
The code

``` ruby
Rake.application.top_level_tasks
```

returns only task names given in a command given. (e.g. when `rails rails:update` is given, only `[rails:update]` is returned, not all tasks in the `rails:` namespace.)

So, I think the below snippet is correct.

``` ruby
namespace :rails do
  task :deprecated do |task|
    Rake.application.top_level_tasks.grep(/^rails:/).each do |task|
      ActiveSupport::Deprecation.warn(<<-MSG.squish)
        Running #{task} with the rails: namespace is deprecated in favor of app.
        Run e.g. bin/rails #{task.sub(/^rails:/, 'app:')} instead."
      MSG
    end
  end

  task :update => ['deprecated', 'app:update']
  task :template => ['deprecated',  'app:template']

  namespace :templates do
    task :copy => ['deprecated', 'app:templates:copy']
  end

  namespace :update do
    task :configs => ['deprecated', 'app:update:configs']
    task :bin => ['deprecated', 'app:update:bin']
  end
end
```

ryohashimoto replied:
OK, I understand.
I removed `Rake.application.top_level_tasks.grep(/^rails:/)`

ryohashimoto commented:
Updated and rebased. I think it's better to use deprecated task names like below.
(And it's not deprecate all tasks with rails: namespace.)

```
%i(update template templates:copy update:configs update:bin).each do |task_name|
  task "rails:#{task_name}" do
    ActiveSupport::Deprecation.warn(<<-MSG.squish)
      Running #{task_name} with the rails: namespace is deprecated in favor of app.
      Run e.g. bin/rails app:#{task_name} instead."
    MSG
    Rake.application.invoke_task("app:#{task_name}")
  end
end
```

ryohashimoto commented:
@kaspth 
Could you take a look?

kaspth commented:
Looks good and thanks for your patience! Can you rebase? If not, I will do it tonight :smile:

kaspth commented:
Wait, what do you mean by "(And it doesn't deprecate all tasks with rails: namespace.)"?

ryohashimoto commented:
@kaspth
Thanks for your support. Rebased now.

> (And it doesn't deprecate all tasks with rails: namespace.)

I mean only tasks listed below are deprecated. If someone add other tasks in the 'rails' namespace,
they wouldn't be deprecated because they don't have corresponding tasks in the 'app' namespace.
I believe this approach is natural.

```` ruby
%i(update template templates:copy update:configs update:bin) # in the rails namespace
````

kaspth commented:
@ryohashimoto thanks!

ryohashimoto commented:
@kaspth Thanks a lot !
